src/main/java/io/netty/incubator/codec/http3/QpackDecoder.java

```diff
@@ -16,94 +16,385 @@
+        sendInsertCountIncrementIfRequired(qpackDecoderStream);
+    }
+
+    /**
+     * Inserts a header field with a literal name corresponding to the
+     * <a href="https://quicwg.org/base-drafts/draft-ietf-quic-qpack.html#name-insert-with-literal-name">
+     *     encoder instruction.</a>
+     *
+     * @param qpackDecoderStream {@link QuicStreamChannel} for the QPACK decoder stream.
+     * @param name of the field.
+     * @param value of the field.
+     * @throws QpackException if the insertion fails.
+     */
+    void insertLiteral(QuicStreamChannel qpackDecoderStream, CharSequence name, CharSequence value)
+            throws QpackException {
+        dynamicTable.add(new QpackHeaderField(name, value));
+        sendInsertCountIncrementIfRequired(qpackDecoderStream);
     }
 
-    private void decodeLiteralWithNameReference(ByteBuf in, BiConsumer<CharSequence, CharSequence> sink)
+    /**
+     * Duplicates a previous entry corresponding to the
+     * <a href="https://quicwg.org/base-drafts/draft-ietf-quic-qpack.html#name-insert-with-literal-name">
+     *     encoder instruction.</a>
+     *
+     * @param qpackDecoderStream {@link QuicStreamChannel} for the QPACK decoder stream.
+     * @param index which is duplicated.
+     * @throws QpackException if duplication fails.
+     */
+    void duplicate(QuicStreamChannel qpackDecoderStream, int index)
             throws QpackException {
-        if ((in.getByte(in.readerIndex()) & 0x10) == 0x10) {
-            final int staticNameIndex = (int) QpackUtil.decodePrefixedInteger(in, 4);
-            assert staticNameIndex >= 0;
-            final CharSequence name = getIndexedName(staticNameIndex);
-            final CharSequence value = decodePrefixedStringLiteral(in);
-            sink.accept(name, value);
+        dynamicTable.add(dynamicTable.getEntryRelativeEncoderInstructions(index));
+        sendInsertCountIncrementIfRequired(qpackDecoderStream);
+    }
+
+    /**
+     * Callback when a bi-directional stream is
+     * <a href="https://quicwg.org/base-drafts/draft-ietf-quic-qpack.html#name-abandonment-of-a-stream"> abandoned</a>
+     *
+     * @param qpackDecoderStream {@link QuicStreamChannel} for the QPACK decoder stream.
+     * @param streamId which is abandoned.
+     */
+    void streamAbandoned(QuicStreamChannel qpackDecoderStream, long streamId) {
+        if (maxTableCapacity == 0) {
+            return;
+        }
+        // https://quicwg.org/base-drafts/draft-ietf-quic-qpack.html#name-stream-cancellation
+        //   0   1   2   3   4   5   6   7
+        // +---+---+---+---+---+---+---+---+
+        // | 0 | 1 |     Stream ID (6+)    |
+        // +---+---+-----------------------+
+        final ByteBuf cancel = qpackDecoderStream.alloc().buffer(8);
+        encodePrefixedInteger(cancel, (byte) 0b0100_0000, 6, streamId);
+        qpackDecoderStream.writeAndFlush(cancel).addListener(CLOSE_ON_FAILURE);
```

Comment:
nit: I think we can do this as a followup but we may want to consider checking if the write was already done / successful when `writeAndFlush(...)` return and so remove some overhead in most cases. Same is true for a lot of other places as well.

Reply:
Good call, I fixed it everywhere.